x-pack/plugins/lens/public/types.ts

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export interface EditorFrameSetup {
+  render: (domElement: Element) => void;
+  registerDatasource: (name: string, datasource: Datasource<any>) => void;
+  registerVisualization: (name: string, visualization: Visualization<any>) => void;
+}
+
+export interface EditorFrameState {
+  visualizations: { [id: string]: object };
+  datasources: { [id: string]: object };
+
+  activeDatasourceId: string;
+  activeVisualizationId: string;
+}
+
+// Hints the default nesting to the data source. 0 is the highest priority
+export type DimensionPriority = 0 | 1 | 2;
+
+// For switching between visualizations and correctly matching columns
+export type DimensionRole =
+  | 'splitChart'
+  | 'series'
+  | 'primary'
+  | 'secondary'
+  | 'color'
+  | 'size'
+  | string; // Some visualizations will use custom names that have other meaning
+
+export interface TableColumns {
+  columnId: string;
+  operation: Operation;
+}
+
+export interface DatasourceSuggestion<T = unknown> {
+  state: T;
+  tableColumns: TableColumns[];
+}
+
+/**
+ * Interface for the datasource registry
+ */
+export interface Datasource<T = unknown> {
+  // For initializing from saved object
+  initialize: (state?: T) => Promise<T>;
+
+  renderDataPanel: (props: DatasourceDataPanelProps) => void;
+
+  toExpression: (state: T) => string;
+
+  getDatasourceSuggestionsForField: (state: T) => Array<DatasourceSuggestion<T>>;
+  getDatasourceSuggestionsFromCurrentState: (state: T) => Array<DatasourceSuggestion<T>>;
+
+  getPublicAPI: (state: T, setState: (newState: T) => void) => DatasourcePublicAPI;
+}
+
+/**
+ * This is an API provided to visualizations by the frame, which calls the publicAPI on the datasource
+ */
+export interface DatasourcePublicAPI {
+  getTableSpec: () => TableSpec;
+  getOperationForColumnId: () => Operation;
+
+  // Called by dimension
+  getDimensionPanelComponent: (
```

Comment:
There was a comment about this but I think it got lost - we don't need the currying with props anymore here because of the getPublicAPI method. Plus `operations: Operation[]` got replaced by `filterOperations` in `DatasourceDimensionPanelProps`. I think this should be `renderDimensionPanelComponent: (props: DatasourceDimensionPanelProps, domElement: Element) => void`